### PR TITLE
Don't shadow special forms

### LINKED ISSUE #60: Should not be able to shadow special forms
So
```
scm> (define define +)
define
scm> (define f 5)
f
```
should work properly, rather than raising an error.

On a related note, `unquote` outside of quasiquotes should raise a different error.

### PR DESCRIPTION
**Pull request by rahularya50**
Fixes #60

### CODE REVIEW

**Comment by rahularya50** [comment]
Please test this thoroughly - I'll do so as well. It passes the automated tests, but I'm very scared of making such a significant modification to the interpreter.

The significant change (other than what the PR is about) is that I now instantiate a special form Callable instance at _lookup time_, rather than during initialization. I don't think this should affect anything, since the Callable instance should never be accessible, but I'm not 100%.

**Comment by rahularya50** [comment]
The following is a brief overview of how lookup of builtins / special forms works, which I'm writing just for my own understanding, but I might as well put it here:

**evaluate_apply.py**
 - Anything that can be called (builtins, procedures, special forms) extends `Callable`.
 - Anything that can be _applied_ (i.e. evaluates all its arguments) extends `Applicable`, which itself extends `Callable`.
 - The frame containing all the builtin procedures (arithmetic, type checking, etc.) is the `Builtins` frame, which is the parent frame of the `Global` frame. The `Builtins` frame is the only frame with no parent frame.

**environment.py**
 - To initialize these frames, two class decorators exist - `global_attr`, and `special_form`. **The `special_form` decorator is new.**  Both decorators take a single argument and put the class they're decorating in a dict. These decorators can be chained.
 - These decorators are both generated by the `make_frame_decorator(defdict)` function, which stores all the decorated functions in `defdict`, which associates a string with each decorated class.
 - All the classes decorated by `global_attr` go in the dictionary `defdict`, while all the classes decorated by `special_form` go in the dictionary `special_forms`.
 - When a new Scheme environment is started, the function `make_global_frame()` is called. It iterates through all the elements of `defdict`, instantiates each class to produce an instance of `Callable`, and puts it in the `Builtins` frame, which is then set to be the parent of `Globals`.

**evaluate_apply.py**
 - When a `Symbol` is evaluated, we just search in the current environment as usual. **This will not return any special forms, since special forms were not added to the ``Builtins`` frame, but only to the dictionary ``special_forms``**.
 **- When a `Pair` is evaluated, we first check if the first element of the `Pair` is a `Symbol`. If so, we call `get_special_form` from `environment.py`, which tries to find the special form class in `special_forms`, and returns a new instance of one if it is found. (otherwise it returns `False`)**
 - Otherwise, we evaluate the first element of the `Pair` recursively as normal.

Before, in `evaluate`, there was a messy way of checking whether a `Symbol` really represents a special form (if it's `Callable` but not `Applicable` or a macro) and faking a `SymbolLookupError` then. This way seems better.

File: editor/environment.py [hunk 1]
```diff
@@ -38,12 +41,19 @@ def __repr__(self):
         return f"#[{self.name}]"
 
 
+def get_special_form(name: str):
+    if name in special_forms:
+        return special_forms[name]()
+    else:
+        return False
```
Comment (kavigupta) [inline editor/environment.py]: Should probably be `None` by python conventions

File: editor/special_forms.py [hunk 2]
```diff
@@ -373,6 +374,12 @@ def quasiquote_evaluate(cls, expr: Expression, frame: Frame, gui_holder: Holder)
             return expr
 
 
+@special_form("unquote")
```
Comment (kavigupta) [inline editor/special_forms.py]: Seems strange to add unquote in this PR

**Review by kavigupta — CHANGES_REQUESTED** [review]

File: editor/special_forms.py [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
+from environment import global_attr, special_form
 from evaluate_apply import Frame, evaluate, Callable, evaluate_all, Applicable
 from log import Holder, VisualExpression, return_symbol, logger
 from helper import pair_to_list, verify_exact_callable_length, verify_min_callable_length, \
     make_list, dotted_pair_to_list
 from lexer import TokenBuffer
 from execution_parser import get_expression
-from scheme_exceptions import OperandDeduceError, IrreversibleOperationError, LoadError, SchemeError
+from scheme_exceptions import OperandDeduceError, IrreversibleOperationError, LoadError, SchemeError, \
+    CallableResolutionError
```
Comment (kavigupta) [inline editor/special_forms.py]: what's this import for?

**Review by kavigupta — CHANGES_REQUESTED** [review]

File: editor/special_forms.py (hunk 3 above)
Comment (rahularya50) [reply editor/special_forms.py]: No idea. Removed.

**Review by kavigupta — APPROVED** [review]

**Pull request merged**